src/Semantic/Api/Terms.hs

```diff
@@ -106,27 +84,71 @@ quietTerm blob = showTiming blob <$> time' ( (doParse blob >>= withSomeTerm (fma
 
-type TermConstraints =
- '[ Taggable
-  , Declarations1
-  , ConstructorName
-  , HasTextElement
-  , Show1
-  , ToJSONFields1
-  , Traversable
-  ]
-
-doParse :: (ParseEffects sig m) => Blob -> m (SomeTerm TermConstraints Loc)
-doParse blob = case blobLanguage blob of
-  Go         -> SomeTerm <$> parse goParser blob
-  Haskell    -> SomeTerm <$> parse haskellParser blob
-  JavaScript -> SomeTerm <$> parse tsxParser blob
-  JSON       -> SomeTerm <$> parse jsonParser blob
-  JSX        -> SomeTerm <$> parse tsxParser blob
-  Markdown   -> SomeTerm <$> parse markdownParser blob
-  Python     -> SomeTerm <$> parse pythonParser blob
-  Ruby       -> SomeTerm <$> parse rubyParser blob
-  TypeScript -> SomeTerm <$> parse typescriptParser blob
-  TSX        -> SomeTerm <$> parse tsxParser blob
-  PHP        -> SomeTerm <$> parse phpParser blob
+
+class ShowTerm term where
+  showTerm :: (Carrier sig m, Member (Reader Config) sig) => term Loc -> m Builder
+
+instance (Functor syntax, Show1 syntax) => ShowTerm (Term syntax) where
+  showTerm = serialize Show . quieterm
+
+instance ShowTerm Py.Term where
+  showTerm = serialize Show . Py.getTerm
+
+
+class SExprTerm term where
+  sexprTerm :: (Carrier sig m, Member (Reader Config) sig) => term Loc -> m Builder
+
+instance (ConstructorName syntax, Foldable syntax, Functor syntax) => SExprTerm (Term syntax) where
+  sexprTerm = serialize (SExpression ByConstructorName)
+
+
+class DOTGraphTerm term where
+  dotGraphTerm :: (Carrier sig m, Member (Reader Config) sig) => term Loc -> m Builder
+
+instance (ConstructorName syntax, Foldable syntax, Functor syntax) => DOTGraphTerm (Term syntax) where
+  dotGraphTerm = serialize (DOT (termStyle "terms")) . renderTreeGraph
+
+
+class JSONTreeTerm term where
+  jsonTreeTerm :: Blob -> term Loc -> Rendering.JSON.JSON "trees" SomeJSON
+
+instance ToJSONFields1 syntax => JSONTreeTerm (Term syntax) where
+  jsonTreeTerm = renderJSONTerm
+
+
+class JSONGraphTerm term where
+  jsonGraphTerm :: Blob -> term Loc -> ParseTreeFileGraph
+
+instance (Foldable syntax, Functor syntax, ConstructorName syntax) => JSONGraphTerm (Term syntax) where
+  jsonGraphTerm blob t
+    = let graph = renderTreeGraph t
+          toEdge (Edge (a, b)) = TermEdge (vertexId a) (vertexId b)
+      in ParseTreeFileGraph path lang (V.fromList (vertexList graph)) (V.fromList (fmap toEdge (edgeList graph))) mempty where
+        path = T.pack $ blobPath blob
+        lang = bridging # blobLanguage blob
+
+
+class (DOTGraphTerm t, JSONGraphTerm t, JSONTreeTerm t, SExprTerm t, ShowTerm t) => TermActions t
+instance (DOTGraphTerm t, JSONGraphTerm t, JSONTreeTerm t, SExprTerm t, ShowTerm t) => TermActions t
+
+doParse
+  :: ( Carrier sig m
+     , Member (Error SomeException) sig
+     , Member Parse sig
+     )
+  => (forall term . TermActions term => term Loc -> m a)
```

Comment:
This function presents something of a problem, and I’m not happy with any of the solutions I’ve landed on thus far.

The problem: given _m_ languages, and _n_ features, fill in the matrix defined by their cartesian product incrementally; or, how do I add support for a language one feature at a time?

Diffing, parsing, analysis, and now symbols all have their own variations on this function, used to allow us to e.g. support `parse --symbols` for precise Python terms now, before we’ve added support for `parse --dot`, `diff`, or anything else. This is to some degree necessary: we don’t support analysis for Markdown or JSON, for example, because they have no computational content. Even so, it’s not without flaws:

- We have several different functions mapping languages to parsers, which increases our surface area for bugs; there’s no one _canonical_ mapping from languages to parsers. If we someday decide JavaScript should be parsed as TypeScript instead of TSX, we’ll have to be very careful to locate and update every place where we’re doing something `doParse`-esque.

- Each such function has to be typechecked independently, and that’s not quick—checking `TermActions` for all of these languages means solving all of its constraints for each `Term` instance, and therefore solving all of _their_ constraints, transitively, for every element of each syntax sum.

- Repeating that effort for each superclass of `TermActions` separately would require multiple `doParse` analogues, which we’d like to avoid due to the aforementioned duplication & margin for error. I have anecdotal evidence suggesting that it’s also slower than grouping them all up under `TermActions` thus.

- While we can easily parameterize `doParse` by the constraint we want to satisfy using `AllowAmbiguousTypes`, `ConstraintKinds`, & `TypeApplications`, doing so moves the burden of solving the constraints to each call-site, slowing down compilation significantly. It also requires us to explicitly mention all of the term types exhaustively in its context, which is tiresome to enumerate.

- Grouping multiple constraints into `TermActions` for improved compile times means that we can’t add support for each member constraint independently of the others; thus, we can’t add `--show` support for Python terms until we’ve also implemented `--dot`, `--json`, etc.

- To the best of my knowledge, there’s no way for us to turn the static satisfiability of a given constraint into a runtime `Maybe` (i.e. to return `Just` for any languages supporting the feature, `Nothing` for the others, using the same function for each different feature).

In an effort to rectify this situation, I have experimented with an approach to allow us to list the parsers for `ShowTerm` separately of the parsers for `JSONTreeTerm`, gaining the compile-time benefits of statically listing all of the supported languages for a given constraint, and the flexibility of tailoring these lists to each feature, without _quite_ so much duplication or compilation time. It’s still pretty duplicative of the map between languages and parsers, tho, and while it doesn’t seem to kill compile times straight away, more experience is needed to know whether it’s viable. I will be working on this in a follow-up PR.

Reply:
This is a wonderful and exhaustive analysis! Many thanks for it, and many thanks for caring about compile times. I don’t have any ideas on what the best way to do this is yet.

Reply:
Fortunately, I’ve made a ton of headway here 👍